### PR TITLE
Declarative language definition

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
* New declarative syntax for languages
    - [Documentation](https://github.com/pokey/cursorless-vscode/blob/language/src/languages/PATTERN_DOCS.md)
* Standardized scope modifier for array and map `items`
* Added scope modifier `regex`
* Support `arg` in plain JavaScript functions
* Support `item` and `map` / `list` for deconstructed lists and objects.
* Support `funk` on default export function
* Support `string` for template literal strings
* Support `name` for function argument identifier
* Improved error message about missing/unimplemented language or scope types.
* Added language support for java
* Added language support for jsx

### CODE REVIEW

File: src/Types.ts [hunk 1]
```diff
@@ -48,21 +48,27 @@ export type ScopeType =
   | "arrowFunction"
   | "class"
   | "className"
+  | "collectionItem"
+  | "collectionKey"
   | "comment"
   | "dictionary"
+  | "xmlElement"
```
Comment (pokey) [inline src/Types.ts:55]: Presumably this applies to react elements as well. Is there a more generic term?

File: src/Types.ts [hunk 2]
```diff
@@ -48,21 +48,27 @@ export type ScopeType =
...
-  | "pair"
-  | "pairKey"
+  | "regex"
   | "statement"
   | "string"
   | "type"
-  | "value";
+  | "value"
+  | "xmlAttribute"
+  | "xmlElement"
```
Comment (pokey) [inline src/Types.ts:68]: Looks like this is duplicate of above?

File: src/Types.ts [hunk 3]
```diff
@@ -282,21 +288,26 @@ export interface DecorationColorSetting {
...
  * Returns the desired relative of the provided node.
  * Returns null if matching node not found.
  **/
-export type NodeFinder = (node: SyntaxNode) => SyntaxNode | null;
+export type NodeFinder = (
+  node: SyntaxNode,
+  selection?: vscode.Selection
+) => SyntaxNode | null;
 
 /** Returns a selection for a given SyntaxNode */
```
Comment (pokey) [inline src/Types.ts:295]: I guess this should now be "one or more selections"?

File: src/languages/java.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  class: "class_declaration",
+  className: "class_declaration[name]",
+  namedFunction: ["method_declaration", "constructor_declaration"],
+  functionName: [
+    "method_declaration.identifier!",
+    "constructor_declaration.identifier!",
+  ],
+  ifStatement: "if_statement",
+  string: "string_literal",
+  name: ["*[declarator.name]", "*[name]", "formal_parameter.identifier!"],
```
Comment (pokey) [inline src/languages/java.ts:54]: What about 
```suggestion
  name: ["*[declarator][name]", "*[name]", "formal_parameter.identifier!"],
```

I think reserving dot for child is clearer

**Review by pokey — COMMENTED** [review]
Just had very quick look and left some comments. Will have another look later. Looks like the new syntax is a win!

File: src/Types.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/Types.ts:55]: Yes because jsx is like html a super set of xml so I think the name is fitting.

File: src/languages/java.ts (hunk 4 above)
Comment (AndreasArvidsson) [reply src/languages/java.ts:54]: But it is a child. Is(name) the child node of the declarator node

Comment (pokey) [reply src/languages/java.ts:54]: Right but "name" is the field name not the child type, as a dot usually indicates, no?

Comment (AndreasArvidsson) [reply src/languages/java.ts:54]: In both cases not just mean a child to a parent. In one case its type in one the case it field names but both are parent child. If you feel strongly about it I can try to fix a regular expression for your format.

Comment (pokey) [reply src/languages/java.ts:54]: So square brackets and dots always mean children, right? But square bracket indicates a field name, and dot indicates a type name. So I'd stick to that convention if that's the case

Comment (AndreasArvidsson) [reply src/languages/java.ts:54]: Correct

Comment (AndreasArvidsson) [reply src/languages/java.ts:54]: Okay get I'll try to get that work

Comment (AndreasArvidsson) [reply src/languages/java.ts]: Fixed!

Comment (pokey) [reply src/languages/java.ts]: Awesome! Looks like it simplified the code a bit too so that's always a good sign

Comment (AndreasArvidsson) [reply src/languages/java.ts]: Yea if you don't count all the time it took to write all those regular expressions ;)

**Pull request merged**